### PR TITLE
First Refactor

### LINKED ISSUE #1: Stricter Checking
Update tsconfig and tslint to use stricter checks on the codebase

### PR DESCRIPTION
**Pull request by aitrean**
Closes
#1 
#5 
#7 
#12
#13 

This is the first refactor for typ3. There are still issues to be addressed from the checklist #20 which will be addressed before initial release.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,104 +1,147 @@
...
+    function bar(bytes32 arg) returns (bool finished) {
+        text = arg;
+        return true;
+    }
 }
+```
+
+You can invoke..
+
+```
```
Comment (HenryNguyen5) [inline README.md]: you can add `ts` at the end of the triple back tick so it highlights the syntax

File: Testing/__tests__/Contract/ContractInstance.spec.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
+import BN from 'bn.js';
+import { ISimpleContract, ISimpleContractConnected, IComplexContract, IComplexContractConnected } from '../../contracts/abiTypes';
+import { ProxiedNode, CreateContract, ContractInstance } from '../../../src/lib';
```
Comment (HenryNguyen5) [inline Testing/__tests__/Contract/ContractInstance.spec.ts:3]: Baseurl is set the src in the tsconfig, you should be able to do `...from lib/`, you might need to include `__tests__` as an included path in the tsconfig though

File: Testing/__tests__/Contract/ContractInstance.spec.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+const complexContractAbi = require('../../contracts/complex/ComplexContract.json')
+const testNode = ProxiedNode(TEST_SERVER_ENDPOINT)
+
+beforeAll(() => startServer());
+afterAll(() => stopServer());
+
+describe('contract instantiation', async () => {
+  const bytesArgument = Buffer.alloc(32, 'testing')
+  const uintArgument = new BN('100000000');
+  const intArgument0 = new BN('-100000000');
```
Comment (HenryNguyen5) [inline Testing/__tests__/Contract/ContractInstance.spec.ts]: unused argument

File: Testing/__tests__/Contract/ContractInstance.spec.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  const intArgument1 = new BN('150000000')
```
Comment (HenryNguyen5) [inline Testing/__tests__/Contract/ContractInstance.spec.ts]: unused argument

File: Testing/contracts/abiTypes.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
+import BN from 'bn.js'
+
+export interface IComplexContract{
```
Comment (HenryNguyen5) [inline Testing/contracts/abiTypes.ts:3]: Got a lot of re-used union types here, you could use a type union assigned to an alias to make it more legible

File: package.json [hunk 6]
```diff
@@ -24,13 +25,38 @@
   },
   "homepage": "https://github.com/Mike-Stupich/typed3#readme",
```
Comment (HenryNguyen5) [inline package.json]: Wrong homepage

File: package.json [hunk 7]
```diff
@@ -24,13 +25,38 @@
...
   "dependencies": {
+    "@types/jest": "22.2.0",
```
Comment (HenryNguyen5) [inline package.json]: Should be in dev dependencies

File: package.json [hunk 8]
```diff
@@ -24,13 +25,38 @@
...
+    "babel-plugin-transform-es2015-modules-commonjs": "^6.26.0",
```
Comment (HenryNguyen5) [inline package.json]: Should be in dev dependencies

Comment (HenryNguyen5) [reply package.json]: Not sure if this is actually needed though

File: package.json [hunk 9]
```diff
@@ -24,13 +25,38 @@
...
     "bn.js": "^4.11.8",
-    "ethereumjs-abi": "^0.6.4",
-    "fetch": "^1.1.0",
-    "node-fetch": "^1.7.3"
+    "ethereumjs-abi": "0.6.4",
+    "ethjs-util": "0.1.4",
+    "fetch": "1.1.0",
+    "ganache-cli": "^6.1.0",
+    "node-fetch": "1.7.3",
+    "npm": "^5.7.1",
```
Comment (HenryNguyen5) [inline package.json]: Uneeded dep

File: package.json [hunk 10]
```diff
@@ -24,13 +25,38 @@
...
+    "rlp": "2.0.0",
+    "source-map-support": "^0.5.4",
```
Comment (HenryNguyen5) [inline package.json]: Should be in dev dependencies

File: package.json [hunk 11]
```diff
@@ -24,13 +25,38 @@
...
+    "timeout-as-promise": "1.0.0",
```
Comment (HenryNguyen5) [inline package.json]: Should be in dev dependencies

File: package.json [hunk 12]
```diff
@@ -24,13 +25,38 @@
...
+    "ts-jest": "22.4.1",
```
Comment (HenryNguyen5) [inline package.json]: Should be in dev dependencies

File: package.json [hunk 13]
```diff
@@ -24,13 +25,38 @@
...
+    "tslint": "^5.9.1",
```
Comment (HenryNguyen5) [inline package.json]: Should be in dev dependencies

File: package.json [hunk 14]
```diff
@@ -24,13 +25,38 @@
...
+    "tslint-config-prettier": "^1.10.0",
```
Comment (HenryNguyen5) [inline package.json]: Should be in dev dependencies

File: package.json [hunk 15]
```diff
@@ -24,13 +25,38 @@
...
+    "types-rlp": "0.0.1",
```
Comment (HenryNguyen5) [inline package.json]: Should be in dev dependencies

File: package.json [hunk 16]
```diff
@@ -24,13 +25,38 @@
...
+    "typescript": "^2.7.2"
```
Comment (HenryNguyen5) [inline package.json]: Should be in dev dependencies

File: src/lib/filter/poll.ts [hunk 17]
```diff
@@ -0,0 +1,8 @@
+import { IProxiedNode } from '../node';
+import delay from 'timeout-as-promise';
```
Comment (HenryNguyen5) [inline src/lib/filter/poll.ts:2]: Dont need this dep, instead use `promisify` from nodejs utils on `setTimeout`

File: src/lib/contract/index.ts [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+
+export const CreateContract = <T>(
+  abi: IAbiBehaviour[],
+  outputMappings: IOutputMappings = {}
+) => {
+  const reducer = (compiledContract: Contract, currMethod: IAbiBehaviour) => {
+    const { name, type } = currMethod
+    const handler = selector[type];
+    switch(type){
+      case(AbiMethodTypes.function):
```
Comment (HenryNguyen5) [inline src/lib/contract/index.ts]: invalid syntax

File: src/lib/contract/index.ts [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+        return {
+          ...compiledContract,
+          [name]: handler(currMethod, outputMappings[name])
+        }
+      case(AbiMethodTypes.constructor):
```
Comment (HenryNguyen5) [inline src/lib/contract/index.ts]: invalid syntax

File: package.json (hunk 8 above)
Comment (aitrean) [reply package.json]: TravisCI build was breaking until I added it. But it worked without otherwise.

**Pull request merged**